R/shared-methods.R

```diff
@@ -509,6 +509,158 @@ DataClass <- R6::R6Class(
         source_data_cols = paste(unlist(self$source_data_cols), collapse = ", ")
       )
       return(sum_df)
+    },
+
+    #' @description Run tests on data class. Inherited by child classes so test
+    #' run through each class.
+    #' @param download logical. To download the data (TRUE) or use a snapshot
+    #' (FALSE). Defaults to FALSE.
+    #' @importFrom purrr walk map
+    #' @importFrom dplyr slice_tail
+    #' @importFrom testthat test_that expect_s3_class expect_true expect_error
+    test = function(download = FALSE) {
+      data_name <- paste0(class(self)[1], " at level ", self$level)
+      raw_path <- paste0(
+        "custom_data/", class(self)[1],
+        "_level_", self$level, ".rds"
+      )
+      if (!file.exists(raw_path)) {
+        download <- TRUE
+      }
+      if (download) {
+        test_that(paste0(data_name, " downloads sucessfully"), {
+          self$download()
```

Comment:
So a potential issue here is that our tests are changing the state of the class. In particular if download is called when using as snapshot. This might be quite unexpected for users and have some strange side effects. There is an argument that perhaps we should be cloning the class and running all tests on that (might be tricky) but open for discussion.

Reply:
One way we could think about this is to split the tests into test and validate where test is testing the infrastructure of the class (downloading etc) and validate is testing that the end data we get is valid in the ways we expect. validate would then only work after all other class methods have been called and so not change state whilst test could change state and include the validate checks.